keras_cv/layers/video_swin_layers.py

```diff
@@ -0,0 +1,980 @@
+        x = self.proj(x)
+
+        if self.norm_layer is not None:
+            x = self.norm(x)
+
+        return x
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "patch_size": self.patch_size,
+                "embed_dim": self.embed_dim,
+            }
+        )
+        return config
+
+
+class VideoSwinPatchMerging(layers.Layer):
+    """Patch Merging Layer in Video Swin Transformer models.
+
+    This layer performs a downsampling step by merging four neighboring patches
+    from the previous layer into a single patch in the output. It achieves this
+    by concatenation and linear projection.
+
+    Args:
+        input_dim (int): Number of input channels in the feature maps.
+        norm_layer (keras.layers, optional): Normalization layer.
+            Default: LayerNormalization
+
+    References:
+        - [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
+        - [Video Swin Transformer GitHub](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """  # noqa: E501
+
+    def __init__(self, input_dim, norm_layer=None, **kwargs):
+        super().__init__(**kwargs)
+        self.input_dim = input_dim
+        self.norm_layer = norm_layer
+
+    def build(self, input_shape):
+        batch_size, depth, height, width, channel = input_shape
+        self.reduction = layers.Dense(2 * self.input_dim, use_bias=False)
+        self.reduction.build(
+            (batch_size, depth, height // 2, width // 2, 4 * channel)
+        )
+
+        if self.norm_layer is not None:
+            self.norm = self.norm_layer(axis=-1, epsilon=1e-5)
+            self.norm.build(
+                (batch_size, depth, height // 2, width // 2, 4 * channel)
+            )
+
+        # compute padding if needed
+        self.pads = [
+            [0, 0],
+            [0, 0],
+            [0, ops.mod(height, 2)],
+            [0, ops.mod(width, 2)],
+            [0, 0],
+        ]
+        self.built = True
+
+    def call(self, x):
+        # padding if needed
+        x = ops.pad(x, self.pads)
+        x0 = x[:, :, 0::2, 0::2, :]  # B D H/2 W/2 C
+        x1 = x[:, :, 1::2, 0::2, :]  # B D H/2 W/2 C
+        x2 = x[:, :, 0::2, 1::2, :]  # B D H/2 W/2 C
+        x3 = x[:, :, 1::2, 1::2, :]  # B D H/2 W/2 C
+        x = ops.concatenate([x0, x1, x2, x3], axis=-1)  # B D H/2 W/2 4*C
+
+        if self.norm_layer is not None:
+            x = self.norm(x)
+
+        x = self.reduction(x)
+        return x
+
+    def compute_output_shape(self, input_shape):
+        batch_size, depth, height, width, _ = input_shape
+        return (batch_size, depth, height // 2, width // 2, 2 * self.input_dim)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "input_dim": self.input_dim,
+            }
+        )
+        return config
+
+
+class VideoSwinWindowAttention(keras.Model):
```

Comment:
```suggestion
class VideoSwinWindowAttention(keras.Layer):
```

Reply:
Subclassing to `keras.Model` instead of layer is on purpose. 

Right now, because of using `keras.Model`,  if we use `model.summary(expand_nested=True)`, it will disclose the sub-layers summary as expected. But not with `keras.Layer`. 

With `keras.Model`

```python
with_model_swin.summary(expand_nested=True)
Model: "new_swin_transformer3d"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 8, 256, 256, 3) │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_patching_and_embeddi… │ (None, 4, 64, 64, 96)  │         9,504 │
│ (VideoSwinPatchingAndEmbedding) │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ embed_proj (Conv3D)        │ ?                      │         9,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ embed_norm                 │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ pos_drop (Dropout)              │ (None, 4, 64, 64, 96)  │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_1         │ (None, 4, 32, 32, 192) │       305,274 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       115,389 │
│ video_swin_transformer_block    │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ identity (Identity)     │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization     │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │        40,797 │
│ video_swin_window_attention     │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense (Dense)        │ ?                      │        27,936 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout (Dropout)    │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_1 (Dense)      │ ?                      │         9,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_1 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_1   │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp (MLP)               │ ?                      │        74,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       115,389 │
│ video_swin_transformer_block_1  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path (DropPath)    │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_2   │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │        40,797 │
│ video_swin_window_attention_1   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_4 (Dense)      │ ?                      │        27,936 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_3 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_5 (Dense)      │ ?                      │         9,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_4 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_3   │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_1 (MLP)             │ ?                      │        74,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ video_swin_patch_merging   │ ?                      │        74,496 │
│ (VideoSwinPatchMerging)         │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_2         │ (None, 4, 16, 16, 384) │     1,200,372 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       451,962 │
│ video_swin_transformer_block_2  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_1 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_5   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       155,322 │
│ video_swin_window_attention_2   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_9 (Dense)      │ ?                      │       111,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_6 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_10 (Dense)     │ ?                      │        37,056 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_7 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_6   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_2 (MLP)             │ ?                      │       295,872 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       451,962 │
│ video_swin_transformer_block_3  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_2 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_7   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       155,322 │
│ video_swin_window_attention_3   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_13 (Dense)     │ ?                      │       111,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_9 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_14 (Dense)     │ ?                      │        37,056 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_10 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_8   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_3 (MLP)             │ ?                      │       295,872 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ video_swin_patch_merging_1 │ ?                      │       296,448 │
│ (VideoSwinPatchMerging)         │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_3         │ (None, 4, 8, 8, 768)   │    11,914,680 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_4  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_3 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_10  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_4   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_18 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_12 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_19 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_13 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_11  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_4 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_5  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_4 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_12  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_5   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_22 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_15 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_23 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_16 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_13  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_5 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_6  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_5 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_14  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_6   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_26 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_18 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_27 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_19 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_15  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_6 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_7  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_6 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_16  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_7   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_30 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_21 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_31 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_22 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_17  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_7 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_8  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_7 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_18  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_8   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_34 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_24 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_35 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_25 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_19  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_8 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_9  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_8 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_20  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_9   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_38 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_27 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_39 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_28 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_21  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_9 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ video_swin_patch_merging_2 │ ?                      │     1,182,720 │
│ (VideoSwinPatchMerging)         │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_4         │ (None, 4, 8, 8, 768)   │    14,232,528 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     7,116,264 │
│ video_swin_transformer_block_10 │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_9 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_23  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │     2,390,760 │
│ video_swin_window_attention_10  │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_43 (Dense)     │ ?                      │     1,771,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_30 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_44 (Dense)     │ ?                      │       590,592 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_31 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_24  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_10 (MLP)            │ ?                      │     4,722,432 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     7,116,264 │
│ video_swin_transformer_block_11 │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_10 (DropPath) │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_25  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │     2,390,760 │
│ video_swin_window_attention_11  │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_47 (Dense)     │ ?                      │     1,771,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_33 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_48 (Dense)     │ ?                      │       590,592 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_34 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_26  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_11 (MLP)            │ ?                      │     4,722,432 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ top_norm (LayerNormalization)   │ (None, 4, 8, 8, 768)   │         1,536 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ avg_pool                        │ (None, 768)            │             0 │
│ (GlobalAveragePooling3D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ classifier (Dense)              │ (None, 400)            │       307,600 │
└─────────────────────────────────┴────────────────────────┴───────────────┘
 Total params: 27,971,494 (106.70 MB)
 Trainable params: 27,971,494 (106.70 MB)
 Non-trainable params: 0 (0.00 B)
```

With `keras.Layer` (same as `expand_nested=False`)


```python
with_layer_swin.summary(expand_nested=True)
Model: "swin_transformer3d_1"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer_1 (InputLayer)      │ (None, 8, 256, 256, 3) │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_patching_and_embeddi… │ (None, 4, 64, 64, 96)  │         9,504 │
│ (VideoSwinPatchingAndEmbedding) │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ pos_drop (Dropout)              │ (None, 4, 64, 64, 96)  │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_1         │ (None, 4, 32, 32, 192) │       305,274 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_2         │ (None, 4, 16, 16, 384) │     1,200,372 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_3         │ (None, 4, 8, 8, 768)   │    11,914,680 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_4         │ (None, 4, 8, 8, 768)   │    14,232,528 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ top_norm (LayerNormalization)   │ (None, 4, 8, 8, 768)   │         1,536 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ avg_pool                        │ (None, 768)            │             0 │
│ (GlobalAveragePooling3D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ classifier (Dense)              │ (None, 400)            │       307,600 │
└─────────────────────────────────┴────────────────────────┴───────────────┘
 Total params: 27,971,494 (106.70 MB)
 Trainable params: 27,971,494 (106.70 MB)
 Non-trainable params: 0 (0.00 B)
```

Some [attempt](https://github.com/keras-team/keras/issues/15250) has been taken in past, but issue remains. Until then, is it okay to keep it as in? Additionally, I wonder why there is `?` in the output shape section?

Reply:
I agree that the summary is much more useful with layers being `keras.Model` subclasses. My suggestion was to maintain consistency across the package. That being said, I do think this might be useful if the standard suite of tests pass with these layers being `keras.Model`s.

About the output shape being `?`: Keras uses the `compute_output_shape` method of the layers to compute the output shape wherever it can. In this case, for example, the `VideoSwinBasicLayer` implements it, so Keras doesn't ever compute the shapes of the layers in the `VideoSwinBasicLayer` model.

Reply:
> My suggestion was to maintain consistency across the package. That being said, I do think this might be useful if the standard suite of tests pass with these layers being keras.Models.

Understood. IMHO, using `keras.Layer` or `keras.Model` don't violate much on consistency. Heck, they're both same, whereas `keras.Model` brings extra stuff, like compile or fit. So, what do you prefer most? Should I update to maintain consistency or keep it as it in? youre call.

Reply:
Let's keep it this way. It seems useful enough without introducing much complexity.